examples/research_projects/codeparrot/scripts/human_eval_mgpu.py

```diff
@@ -0,0 +1,218 @@
+from transformers import (
+    AutoModelForCausalLM,
+    AutoTokenizer,
+    HfArgumentParser,
+    StoppingCriteria,
+    StoppingCriteriaList,
+    set_seed,
+)
+
+
+EOF_STRINGS = ["\nclass", "\ndef", "\n#", "\n@", "\nprint", "\nif"]
+
+
+class TokenizeDataset(IterableDataset):
+    """Tokenize and preprocess the dataset
+    Multiple copies of the same prompt are sent sequentially.
+    See compute_code for more details.
+    """
+    def __init__(self, tokenizer, dataset, n_tasks=None, n_copies=1):
+        self.tokenizer = tokenizer
+        self.dataset = dataset
+        self.n_tasks = len(dataset) if n_tasks is None else n_tasks
+        self.n_copies = n_copies
+
+    def __iter__(self):
+        for task in range(self.n_tasks):
+            # without strip, the model generate commented codes ...
+            prompt = self.tokenizer.eos_token + self.dataset[task]['prompt'].strip()
+            # codeparrot model is not robust to padding
+            input_ids = self.tokenizer(prompt)['input_ids']
+            for _ in range(self.n_copies):
+                yield {"ids": torch.tensor(input_ids)}
+
+
+class EndOfFunctionCriteria(StoppingCriteria):
+    """Custom `StoppingCriteria` which checks if all generated functions in the batch are completed."""
+
+    def __init__(self, start_length, eof_strings, tokenizer):
+        self.start_length = start_length
+        self.eof_strings = eof_strings
+        self.tokenizer = tokenizer
+
+    def __call__(self, input_ids, scores, **kwargs):
+        """Returns true if all generated sequences contain any of the end-of-function strings."""
+        decoded_generations = self.tokenizer.batch_decode(input_ids[:, self.start_length:])
+        done = []
+        for decoded_generation in decoded_generations:
+            done.append(any([stop_string in decoded_generation for stop_string in self.eof_strings]))
+        return all(done)
+
+
+def remove_last_block(string):
+    """Remove the last block of the code containing EOF_STRINGS
+    """
+    string_list = re.split("(%s)" % "|".join(EOF_STRINGS), string)
+    # last string should be ""
+    return "".join(string_list[:-2])
+
+
+def complete_code(accelerator, model, tokenizer, dataloader, n_tasks, batch_size=20, **gen_kwargs):
+    """Generate multiple codes for each task in the dataset. This function leverage accelerator to distribute
+    the processing to multiple GPUs.
+    dataloader, a wrapper around a TokenizeDataset objectm is supposed to send all the prompts from 
+    the evalution dataset to the modelm as the following:
+    [p_0_0, p_0_1, ..., p_0_nc-1, p_1_0, ..., p_nt-1_nc-1]
+    where nc is the number of copies of the prompt, and nt is the number of tasks.
+    nc is such that num_sample = nc * batch_size
+    In this way, the same prompt is sent to the model on all the GPUs to avoid zero padding.
+    (accelerate require all the batch having the same shape before feeding it to multiple GPUs)
```

Comment:
Do I understand this correctly: if we want 100 generations and have 4 GPUs then each GPU will get 25 times the same sample. However, if we have 8 GPUs this will break, as 100 is not divisible by 8, right? As we scale to more (and larger) models I suspect that we will likely scale the GPUs and so that would be quite a limiting factor.

Reply:
Yes, this is the idea